File: dp/house_robber_ii.py
"""
Note: This is an extension of House Robber.

After robbing those houses on that street, the thief has found himself a new

place for his thievery so that he will not get too much attention. This time,

all houses at this place are arranged in a circle. That means the first house

is the neighbor of the last one. Meanwhile, the security system for these houses

remain the same as for those in the previous street.

********************************************************************************

Given a list of non-negative integers representing the amount of money of each

house, determine the maximum amount of money you can rob tonight without alerting the police.


"""
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def rob(self, nums):
        n = len(nums)
        if n == 0: return 0
        if n <= 2: return max(nums)
        val1 = nums[0]
        val2 = max(nums[0], nums[1])

        for i in xrange(2, n):
            # which is larger?
            # num[i] + num[i-2] or num[i-1]
            val1, val2 = val2, max(val2, val1 + nums[i])

        kmax = val2
        nums += nums
        val1 = nums[0]
        val2 = max(nums[0], nums[1])
        for i in xrange(2, 2*n):
            # which is larger?
            # num[i] + num[i-2] or num[i-1]
            val1, val2 = val2, max(val2, val1 + nums[i])

        return val2 - kmax


sol = Solution()
print sol.rob([1,2,3,4,5])

File: various/ruby/intermediate/house_robber_ii.rb
# https://leetcode.com/problems/house-robber-ii
# # @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  return 0 if nums.size == 0
  return nums.max if nums.size <= 2
  # rob from 0 to nums.size-2
  res1 = normal_rob(nums[0..-2])
  # rob from 1 to nums.size-1
  res2 = normal_rob(nums[1..-1])

  [res1, res2].max
end

def normal_rob(nums)
  dp = Array.new(nums.size, 0)
  dp[0] = nums[0]
  dp[1] = [dp[0], nums[1]].max
  for i in 2..nums.size-1
    dp[i] = [dp[i-2] + nums[i], dp[i-1]].max
  end

  dp[-1]
end

p rob([1,2,3,1])
p rob([2,3,2])
